src/backend/distributed/transaction/transaction_management.c

```diff
@@ -423,3 +435,48 @@ ActiveSubXacts(void)
 
 	return activeSubXactsReversed;
 }
+
+
+/*
+ * If an ERROR is thrown while processing a transaction the ABORT handler is called.
+ * ERRORS thrown during ABORT are not treated any differently, the ABORT handler is also
+ * called during processing of those. If an ERROR was raised the first time through it's
+ * unlikely that the second try will succeed; more likely that an ERROR will be thrown
+ * again. This loop continues until Postgres notices and PANICs, complaining about a stack
+ * overflow.
+ *
+ * Instead of looping and crashing, SwallowErrors lets us attempt to continue running the
+ * ABORT logic. This wouldn't be safe in most other parts of the codebase, in
+ * approximately none of the places where we emit ERROR do we first clean up after
+ * ourselves! It's fine inside the ABORT handler though; Postgres is going to clean
+ * everything up before control passes back to us.
+ */
+static void
+SwallowErrors(void (*func)())
+{
+	MemoryContext savedContext = CurrentMemoryContext;
+
+	PG_TRY();
+	{
+		func();
```

Comment:
Assuming `func` can also do database stuff, we should start a subtransaction before `PG_TRY()` and rollback to it in `PG_CATCH()` (and release it at the end of `PG_TRY()`) so we also restore transaction state. See [CollectBasicUsageStatistics()](https://github.com/citusdata/citus/blob/release-7.3/src/backend/distributed/utils/statistics_collection.c#L104).

It doesn't do database access in the single usage in this PR, but assuming `SwallowErrors()` might be used in future, it is safer to do it.

Reply:
This meant to be called from the abort handler, so it's probably too late to open a subtransaction.